chore: add dictionary set, get, delete

poppoerika opened the pull request:
Added dictionary set, get, and delete APIs and tests to check those operations.

Ticket: #17

pgautier404 commented on src/Cache/_ScsDataClient.php:
```diff
@@ -114,25 +114,43 @@ public function set(string $cacheName, string $key, string $value, int $ttlSecon
     }
 
-    public function get(string $cacheName, string $key) : CacheGetResponse
+    private function ttlToMillis(?int $ttl = null): int
+    {
+        if (!$ttl) {
+            $ttl = $this->defaultTtlSeconds;
+        }
+        validateTtl($ttl);
+        return $ttl * 1000;
+    }
+
+    private function processCall(UnaryCall $call): mixed
+    {
+        [$response, $status] = $call->wait();
+        if ($status->code !== 0) {
+            throw _ErrorConverter::convert($status->code, $status->details, $call->getMetadata());
+        }
+        return $response;
+    }
```
Can you move these definitions up so they live between `__construct()` and `set()` please?

pgautier404 commented on src/Cache/_ScsDataClient.php:
```diff
@@ -354,14 +369,92 @@ public function listErase(string $cacheName, string $listName, ?int $beginIndex=
+            return new CacheDictionaryGetResponseMiss();
+        }
+        if ($dictionaryGetResponse->getFound()->getItems()->count() == 0) {
+            return new CacheDictionaryGetResponseError(new Exception("_DictionaryGetResponseResponse contained no data but was found"));
```
The exception used to construct the Error response class needs to be an SdkError like the one above this on line 428. Please use `UnknownError` like you did above unless you think one of the other SdkError-derived classes works better.

pgautier404 commented on tests/Cache/CacheClientTest.php:
```diff
@@ -117,26 +116,40 @@ public function testCreateCacheEmptyName()
         $this->assertEquals(MomentoErrorCode::AUTHENTICATION_ERROR, $response->asError()->errorCode());
     }
 
+    private function getBadAuthTokenClient(): SimpleCacheClient
+    {
+        $badEnvName = "_MOMENTO_BAD_AUTH_TOKEN";
+        putenv("{$badEnvName}={$this->BAD_AUTH_TOKEN}");
+        $authProvider = new EnvMomentoTokenProvider($badEnvName);
+        putenv($badEnvName);
+        return new SimpleCacheClient($authProvider, $this->DEFAULT_TTL_SECONDS);
+    }
+
```
Can you move this to the top of the class, please, either before `setUp()` or after `tearDown()`?

pgautier404 commented on tests/Cache/CacheClientTest.php:
```diff
@@ -77,13 +67,18 @@ public function testCreateSetGetDelete() {
 
-    // Client initialization tests
-    public function testNegativeDefaultTtl() {
+    // Happy path test
```
Please move this comment up so it appears before the `testCreateSetGetDelete` test.

pgautier404 commented on tests/Cache/CacheClientTest.php:
```diff
@@ -92,22 +87,26 @@ public function testNonJwtTokens() {
         $this->assertNotNull($response->asAlreadyExists());
     }
 
+    // Create cache tests
```
This comment should come before the `testCreateCacheAlreadyExists` test please.

poppoerika commented:
@pgautier404 Apologize for removing the list tests you've added. 
I need to be careful when I merge and use PhpStorm interface to manage these merges 🙇‍♀️ 
I addressed your feedback in this push.

pgautier404 reviewed (dismissed):
LGTM!

malandis commented on src/Cache/_ScsDataClient.php:
```diff
@@ -354,14 +369,92 @@ public function listErase(string $cacheName, string $listName, ?int $beginIndex=
+            $dictionarySetRequest->setItems([$this->toSingletonFieldValuePair($field, $value)]);
+            $dictionarySetRequest->setRefreshTtl($refreshTtl);
+            $dictionarySetRequest->setTtlMilliseconds($ttlMillis);
+            $call = $this->grpcManager->client->DictionarySet($dictionarySetRequest, ["cache" => [$cacheName]], ["timeout" => $this->deadline_seconds * self::$TIMEOUT_MULTIPLIER]);
```
I recommend pre-computing this so we don't multiply this out in all the functions, ie in the constructor store `deadline_microseconds = $this->deadline_seconds * self::$TIMEOUT_MULTIPLIER` and use that here

pgautier404 replied:
I'd rather avoid using `deadline_microseconds` because they're technically not microseconds. I think this is a great idea, but I'd like to think about it a bit . . . I'll make this change in an upcoming PR.

malandis commented on tests/Cache/CacheClientTest.php:
```diff
@@ -768,4 +805,129 @@ public function testListErase_MissingList()
+        $this->assertEquals($value, $response->asHit()->value());
+    }
+
+    public function testDictionaryDeleteThrowExceptionForEmptyCacheName()
```
function name is now out of date -- returns error instead of throws exception :)

malandis reviewed (commented):
Few things:
- Recommend storing the deadline in microseconds to avoid duplicated code
- Recommend splitting data structures tests into multiple files, one for each data structure
- Consistent test name format (no underscores)

poppoerika commented:
@malandis Thank you for the feedback!
Regarding splitting the tests, we'll address that in a follow-up PR. (https://github.com/momentohq/client-sdk-php/issues/22)
I will update the test names though.

poppoerika commented:
@malandis We will address test names in a follow-up pr to get this pr merge first.
